R/env.R

```diff
@@ -0,0 +1,702 @@
+#' expression sees when it is evaluated. In R, scope is hierarchical:
+#' each environment is defined with a parent environment. An
+#' environment and its grand-parents form together a linear
+#' hierarchy. All objects within the grand-parents are in scope unless
+#' they are eclipsed by synonyms (other bindings with the same names)
+#' in child environments.
+#'
+#' \code{env()} is a s3 generic. Methods are provided for functions,
+#' formulas and frames. If called with a missing argument, the
+#' environment of the current evaluation frame (see
+#' \code{\link{eval_stack}()}) is returned. If you call \code{env()}
+#' with an environment, it acts as the identity function and the
+#' environment is simply returned (this helps simplifying code when
+#' writing generic functions).
+#'
+#' \code{env_new()} creates a new environment. \code{env_next()}
+#' returns the parent environment of \code{env} if called with \code{n
+#' = 1}, the grand-parent with \code{n = 2}, etc. \code{env_tail()}
+#' searches through the parents and returns the one which has
+#' \code{\link{env_empty}()} as parent.
+#'
+#' @param env An environment or an object with a S3 method for
+#'   \code{env()}. If missing, the environment of the current
+#'   evaluation frame is returned.
+#' @param parent A parent environment. Can be an object with a S3
+#'   method for \code{env()}.
+#' @param dict A vector with unique names which defines bindings
+#'   (pairs of name and value). See \code{\link{is_dict}()}.
+#' @param n The number of generations to go through.
+#' @seealso \link{env_scoped}, \code{\link{env_has}()},
+#'   \code{\link{env_assign}()}.
+#' @export
+#' @examples
+#' # Get the environment of frame objects. If no argument is supplied,
+#' # the current frame is used:
+#' fn <- function() {
+#'   list(
+#'     env(call_frame()),
+#'     env()
+#'   )
+#' }
+#' fn()
+#'
+#' # Environment of closure functions:
+#' env(fn)
+#'
+#' # There is also an assignment operator:
+#' env(fn) <- env_base()
+#' env(fn)
+#'
+#'
+#' # env_new() creates by default an environment whose parent is the
+#' # current environment. Here we return a new environment that has
+#' # the evaluation environment (or frame environment) of a function
+#' # as parent:
+#' fn <- function() {
+#'   my_object <- "A"
+#'   env_new()
+#' }
+#' frame_env <- fn()
+#'
+#' # The new environment is empty:
+#' env_has(frame_env, "my_object")
+#'
+#' # But sees objects defined inside fn() by inheriting from its
+#' # parent:
+#' env_sees(frame_env, "my_object")
+#'
+#'
+#' # Create a new environment with a particular scope by setting a
+#' # parent. When inheriting from the empty environment, the
+#' # environment will have no object in scope at all:
+#' env <- env_new(env_empty())
+#' env_sees(env, "lapply")
+#'
+#' # The base package environment is often a good default choice for a
+#' # parent environment because it contains all standard base
+#' # functions:
```

Comment:
But note that this will never inherit from packages the user has loaded (and maybe add a note about why it's impossible to get that?)